### ISSUE TITLE
Tenants

### ISSUE THREAD

**Issue by samuel-utbult-oborgen** [HoudiniGraphql/houdini#637]
### Describe the feature

I would like to use Houdini in the project I am currently working on. The system with the GraphQL API the client would connect to have different tenants. A client must set the header `X-Tenant` with the ID of the tenant to use a specific tenant when making a request. The tenant the user is using is decided by the URL, as each path is prefixed with `/tenant/[tenantID]`. This works very nicely with SvelteKit, as every page simply is located under `/src/tenant/[tenantID]`.

However, I do not see any obvious way to pass the tenant ID to the fetchQuery function in client.js. I have thought about a few workarounds, but everyone has failed since certain users have access to more than one tenant and they need to be able to switch between tenants within the SPA. This creates a case where Houdini’s built-in caching need to be aware of what tenant the users are requesting data for so that responses can be cached separately for each tenant.

One method I have tried is to get the tenant from the URL in the Variables function and set it as a variable. However, this would mean that some logic needs to be implemented to set the tenant in each GraphQL query that is made, which is repetitive. Additionally, not all queries will have variables, which would mean variables need to be added unnecessarily. Also, in principle, a tenant ID is not really a variable as it is not passed as a variable to the GraphQL query on the server but handled using different logic.

To me, metadata looks like a good match for a location where the tenant could be located. However, I am using the standard declarative form of defining that a page requests a query, i.e:

```
import type { PageData } from "./$houdini";
export let data: PageData;
 $: ({ GetEntities } = data);

// When using the data in the view.
$GetEntities.data?.entities
```

You have a guide on your website where you describe that metadata can be used when using the more imperative `.fetch()` method but I cannot find any information about the standard declarative form. I have researched your source code to some extent and to me, it looks like this is currently not possible. 

Also, does the caching mechanism regard metadata? Because, if not, then using metadata to specify the tenant would not be possible.

Preferably, I would like an API where it would be possible to add the tenant ID to the metadata in a load function located in `/src/tenant/[tenantID]/+page.ts` as this would mean the tenant would be cascaded to all pages.

As a plan B, would it be otherwise possible to have a middleware where a variable could be injected? This would preferably be in `/src/tenant/[tenantID]/+page.ts` just as with the case described above. However, a global middleware would be fine if it is invoked each page change not just on the server but also on the client as well. It would be okay for me with such a solution even though a tenant ID is not a variable in principle.

I am going to continue investigating your source code and might propose a solution if I find one.

### Criticality

waiting for it to switch all my project to Houdini

**Comment by samuel-utbult-oborgen** [comment]
I researched the code further and found out that only the variables are used when caching and not the metadata. Thus, it seems like some kind of variable middleware would be needed.

**Comment by AlecAivazis** [comment]
Oh, very interesting 🤔 I'll have to think a little bit about how to support this. You're correct that there isn't an immediately obvious solution. 

Are the IDs globally unique across tenants? Once we _do_ have a way to point your queries to the right API, we need to make sure that the IDs are unique so the cache doesn't get confused.

**Comment by samuel-utbult-oborgen** [comment]
The IDs for the entities belonging to a certain tenant are globally unique across tenants in our backend implementation, as they refer to database keys of rows that also each have a foreign key to a tenant. However, I need to support operations that does not include a key, such as listing all entries of a certain entity within the tenant (with or without pagination). In this case, no ID is included as a query variable. Additionally, a popular alternative to the described tenant implementation is to have one database per tenant, which means the entry IDs would not be unique.

I researched the code further and saw that logic exists to only cache variables that are used by the query, so even though I have managed to inject the tenant ID in the variables, it won’t be cached. Yet, as I’ve understood it, we users can pass any arbitrary data as metadata, so if the metadata is cached and can be provided from /src/tenant/[tenantID]/+page.ts, it could be a good match for specifying a tenant ID.

I am going to investigate how metadata could be cached today. I would like to propose that two callbacks could be injected: one to add or modify the metadata and one to compare two metadata objects so that the caching mechanism could differentiate between two metadata instances and cache the results separately. Additionally, since probably only a subset of the library’s users will use this functionality, I propose that metadata caching only would be done if a comparator callback has been provided. What do you think about this?

**Comment by samuel-utbult-oborgen** [comment]
Sorry, I just realized what you are talking about. Since you cache the result on an entry basis, errors would be introduced if you just looked at the IDs of the individual entries without regarding the tenant. Simply “caching the metadata” as I proposed would not work here. In my case, this would not be an issue since the IDs are unique anyway, but it would not work if the tenants were stored in different databases. I guess each entry would have to be identified using a pair containing the tenant ID and the entity ID.

Nevertheless, I think being able to inject metadata when using the declarative form would still be a good idea, as it would mean the metadata feature would be accessible with this type of syntax and not just from `.fetch()`.

**Comment by samuel-utbult-oborgen** [comment]
I have managed to create a solution where it’s possible to inject variables through a middleware while ensuring that these extra variables are included in the cache key. You can see it at this link: https://github.com/oborgen/houdini Only packages/houdini-svelte/src/runtime/stores/query.ts and packages/houdini/src/runtime/cache/stuff.ts have been updated. Notice that entries with the same ID in different tenants would probably not work.

**Comment by samuel-utbult-oborgen** [comment]
@AlecAivazis Have you thought anything more about this? I have a proposal: The file `packages/houdini/src/runtime/cache/index.ts` currently contains a global instance of Cache. This could be replaced with an object that maps tenants (as strings) to Cache instances. Then, getCache() would have to accept a tenant, which it could use to look up the correct Cache instance. This would ensure isolation between tenants from what I can see. However, each time anything is queried, mutated, subscribed to, etc, the tenant would have to be specified (with a default value of course as most users will not have to deal with tenants).

It would be quite nice to have some way to set the tenant once (based on what page is loaded, etc) instead of having to set it on each request. One way this might be possible would be to use Svelte's setContext/getContext, which would mean users could set the tenant with setContext when a page is loading and then getCache could get it using getContext so that the right cache could be returned. However, an immediate roadblock to this is that LayoutSessionStore.getConfig calls getCache outside of any component initialization, which causes an exception to be thrown when calling getContext. The configuration could probably be something the Cache instances could share but I haven't investigated how that currently works.

Another potential problem with setContext/getContext might be that many stores seem to be created as global instances, which might cause problems. However, to me, most of them seem to be bound to specific pages, which possibly could make it so the correct tenants would be returned from getContext. What’s your opinion on this?

**Comment by AlecAivazis** [comment]
Sorry i haven't reponded here @samuel-utbult-oborgen. I've been super busy with work and the `1.0` release. My overall feeling is that your situation is very specific and it's hard to imagine that there would be a universal way to address this as part of houdini's core. 

I think the best approach would be to put together a custom plugin that does whatever is necessary. One thing you could explore is having a plugin that changes the logic of the generated runtime. That would let you customize the behavior to fit your exact needs. Let me know if that sounds like something you want to explore and we can talk through the options/steps.

**Comment by samuel-utbult-oborgen** [comment]
It's fine. I have been busy with other things as well. My current workaround is to put `data-sveltekit-reload` on any link that changes which tenant the user is currently using. This effectively means the cache is cleared when switching tenants. As you write, some kind of plugin could be useful as long as it could overwrite the behavior of getCache to maintain separate cache instances for the different tenants. Do you have any plans to introduce support for plugins to the library? I am not a Javascript-expert, but as far as I understand, patterns like monkey-patching can be used to interfere with the behavior of existing functions. I can look into that when I have time.

**Comment by AlecAivazis** [comment]
> Do you have any plans to introduce support for plugins to the library?

Yes! In fact, the svelte bindings are one such plugin to the core houdini infrastructure. There are already a few people in the wild who have custom plugins to fit their needs so it's definitely doable. Only big hurdle is that it's not really documented yet since I want to reserve the ability to change it but if this is something you want to explore I can point you in the right direction (or we can chat on discord if that works better for you).

**Comment by jycouet** [comment]
In https://github.com/HoudiniGraphql/houdini/tree/main/packages, you can already see a few plugin examples: 
- [houdini-svelte](https://github.com/HoudiniGraphql/houdini/tree/main/packages/houdini-svelte)
- [houdini-plugin-svelte-global-stores](https://github.com/HoudiniGraphql/houdini/tree/main/packages/houdini-plugin-svelte-global-stores) _(not released yet)_

**Comment by samuel-utbult-oborgen** [comment]
Thank you. I can look into how those plugins work to see how a tenants plugin could be implemented.

**Comment by AlecAivazis** [comment]
Awesome! For a better overview of all of the hooks that a plugin can implement, here is a diagram that should be up to date: https://app.excalidraw.com/l/5mJ4nb9Inaz/7QsnYkLMbK3
